[UnifiedPDF] UnifiedPDFPlugin should be able to query PDFKit for annotations under a page point

aprotyas opened the pull request:
#### b3ccba3d047d8f956609df0364fd5b7b95a71efa
<pre>
[UnifiedPDF] UnifiedPDFPlugin should be able to query PDFKit for annotations under a page point
<a href="https://bugs.webkit.org/show_bug.cgi?id=265810">https://bugs.webkit.org/show_bug.cgi?id=265810</a>
<a href="https://rdar.apple.com/119147700">rdar://119147700</a>

Reviewed by Tim Horton.

We currently don&apos;t perform the required transformations to map a point
from the root view to the page space, and as such cannot perform
hit-testing to get the PDFAnnotation objects of interest under a page
point.

This commit introduces the required coordinate space plumbing so that
UnifiedPDFPlugin is able to reason in the PDFPage space. This allows us
to query PDFPage for annotations under a given point, which we need in
anticipation of getting cursor updates to work correctly over text and
other annotation types.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.h:
(WebKit::PDFDocumentLayout::displayMode const):

Make displayMode publicly accessible.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.mm:
(WebKit::PDFDocumentLayout::updateLayout):
(WebKit::PDFDocumentLayout::layoutSingleColumn):
(WebKit::PDFDocumentLayout::layoutTwoUpColumn):
(WebKit::PDFDocumentLayout::documentMargin): Deleted.
(WebKit::PDFDocumentLayout::pageMargin): Deleted.

Move these to just be static constexpr properties on PDFDocumentLayout,
and adopt the change at call-sites.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::convertFromPluginToDocument const):
(WebKit::UnifiedPDFPlugin::nearestPageIndexForDocumentPoint const):
(WebKit::UnifiedPDFPlugin::convertFromDocumentToPage const):
(WebKit::UnifiedPDFPlugin::pdfElementTypesForPluginPoint const):

Introduce functions that return the data reflected by their names, all
to be consumed by pdfElementTypesForPluginPoint in anticipation of
cursor updates.

(WebKit::UnifiedPDFPlugin::performContextMenuAction const):
(WebKit::UnifiedPDFPlugin::handleContextMenuEvent):

Drive-by fix, make `#endif` scopes clearer.

Canonical link: <a href="https://commits.webkit.org/271528@main">https://commits.webkit.org/271528@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/af10ddcd0e5fdcb907e4c6dc388e54411872b172

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28683 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7326 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30054 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31304 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/26175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9443 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4687 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26249 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28953 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6057 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24652 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5263 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/5416 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/32642 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26257 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26099 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/31658 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/5380 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/3551 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/29436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6995 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6860 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/5847 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/5898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -150,6 +150,9 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
 
     RefPtr<WebCore::GraphicsLayer> createGraphicsLayer(const String& name, GraphicsLayer::Type);
 
+    WebCore::IntPoint pointInDocumentSpaceForPluginPoint(const WebCore::IntPoint&) const;
```
I think these should use standard WebKit naming (like LocalFrameView, and like the existing ones on PDFPluginBase): convertFromXToY(), and _not_ include the type ("point")

aprotyas replied:
Made names more standard in [af10ddc](https://github.com/WebKit/WebKit/commit/af10ddcd0e5fdcb907e4c6dc388e54411872b172)

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -460,9 +464,88 @@
+        notImplemented();
+        break;
+    case PDFDocumentLayout::DisplayMode::SinglePage:
+        // There is only one page to be near.
```
I think smfr's plan for SinglePage is that they're actually all laid out in a row and we just use scroll-snap, so it's likely the Continuous code is the right way to go for this. (see also the comment in `PDFDocumentLayout::layoutPages`)

aprotyas replied:
Interpreted too literally -- will fallthrough on the `SinglePage` case now.

aprotyas replied:
Fixed in [af10ddc](https://github.com/WebKit/WebKit/commit/af10ddcd0e5fdcb907e4c6dc388e54411872b172)

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -460,9 +464,88 @@
+
+    if (auto annotation = [m_documentLayout.pageAtIndex(nearestPageIndex) annotationAtPoint:pointInPDFPageSpace]) {
+        // FIXME: Reason about the annotation type to construct the full set of PDFElementTypes at this page point.
+        (void)annotation;
```
UNUSED_PARAM, perhaps

aprotyas replied:
Fixed in [af10ddc](https://github.com/WebKit/WebKit/commit/af10ddcd0e5fdcb907e4c6dc388e54411872b172)

sammygill commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.mm:
```diff
@@ -170,8 +160,8 @@
-    auto documentMargin = PDFDocumentLayout::documentMargin();
-    auto pageMargin = PDFDocumentLayout::pageMargin();
+    auto documentMargin = PDFDocumentLayout::documentMargin;
+    auto pageMargin = PDFDocumentLayout::pageMargin;
```
We could probably consider getting rid of the local variables now if we wanted to

aprotyas replied:
Fixed in [af10ddc](https://github.com/WebKit/WebKit/commit/af10ddcd0e5fdcb907e4c6dc388e54411872b172)

sammygill commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -460,9 +464,88 @@
+            // nearest page to the point (so far), so we always keep track of it.
+            nearestPageIndex = index;
+            if (point.y() <= pageBounds.maxY())
+                break;
```
Is it possible to ever *not* hit this condition? If so `nearestPageIndex` would point to the last page and would that be correct in that scenario or would `std::nullopt` be more appropriate?

aprotyas replied:
Yes, it is possible when the point is after the last page, but at any rate the last page would be closest to that point.

It doesn't matter too much because PDFPage wouldn't return an annotation for an out-of-bound point anyways, but yes I'll change the behavior here to return `std::nullopt` if we don't hit this condition at all.

aprotyas replied:
Addressed in [af10ddc](https://github.com/WebKit/WebKit/commit/af10ddcd0e5fdcb907e4c6dc388e54411872b172)

aprotyas committed:
[UnifiedPDF] UnifiedPDFPlugin should be able to query PDFKit for annotations under a page point
https://bugs.webkit.org/show_bug.cgi?id=265810
rdar://119147700

Reviewed by Tim Horton.

We currently don't perform the required transformations to map a point
from the root view to the page space, and as such cannot perform
hit-testing to get the PDFAnnotation objects of interest under a page
point.

This commit introduces the required coordinate space plumbing so that
UnifiedPDFPlugin is able to reason in the PDFPage space. This allows us
to query PDFPage for annotations under a given point, which we need in
anticipation of getting cursor updates to work correctly over text and
other annotation types.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.h:
(WebKit::PDFDocumentLayout::displayMode const):

Make displayMode publicly accessible.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDocumentLayout.mm:
(WebKit::PDFDocumentLayout::updateLayout):
(WebKit::PDFDocumentLayout::layoutSingleColumn):
(WebKit::PDFDocumentLayout::layoutTwoUpColumn):
(WebKit::PDFDocumentLayout::documentMargin): Deleted.
(WebKit::PDFDocumentLayout::pageMargin): Deleted.

Move these to just be static constexpr properties on PDFDocumentLayout,
and adopt the change at call-sites.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::convertFromPluginToDocument const):
(WebKit::UnifiedPDFPlugin::nearestPageIndexForDocumentPoint const):
(WebKit::UnifiedPDFPlugin::convertFromDocumentToPage const):
(WebKit::UnifiedPDFPlugin::pdfElementTypesForPluginPoint const):

Introduce functions that return the data reflected by their names, all
to be consumed by pdfElementTypesForPluginPoint in anticipation of
cursor updates.

(WebKit::UnifiedPDFPlugin::performContextMenuAction const):
(WebKit::UnifiedPDFPlugin::handleContextMenuEvent):

Drive-by fix, make `#endif` scopes clearer.

Canonical link: https://commits.webkit.org/271528@main
